cmd/kubeadm/app/preflight/checks.go

```diff
@@ -841,6 +842,88 @@ func (ResolveCheck) Check() (warnings, errors []error) {
 	return warnings, errors
 }
 
+const (
+	// IPVSProxyMode is match set up cluster with ipvs proxy model
+	IPVSProxyMode = "ipvs"
+)
+
+// Sets of IPVS required kernel modules.
+var ipvsModules = []string{
+	"ip_vs",
+	"ip_vs_rr",
+	"ip_vs_wrr",
+	"ip_vs_sh",
+	"nf_conntrack_ipv4",
+}
+
+// KernelModuleIPVSRequiredCheck tests IPVS required kernel moduler.
+type KernelModuleIPVSRequiredCheck struct{}
+
+// Name returns label for KernelModuleIPVSRequiredCheck
+func (KernelModuleIPVSRequiredCheck) Name() string {
+	return "KernelModuleIPVSRequired"
+}
+
+// LinuxKernelHandler implements KernelHandler interface.
+type LinuxKernelHandler struct {
+	executor utilsexec.Interface
+}
+
+// Check try to validates IPVS required kernel modules exists or not.
+func (KernelModuleIPVSRequiredCheck) Check() (warnings, errors []error) {
+	glog.V(1).Infoln("validating the kernel module IPVS required exists in Master/Node or not")
+
+	warnings = []error{}
+	// Find out loaded kernel modules
+	handle := &LinuxKernelHandler{}
+	out, err := handle.executor.Command("cut", "-f1", "-d", " ", "/proc/modules").CombinedOutput()
+	if err != nil {
+		errors = append(errors, fmt.Errorf("error getting installed ipvs required kernel modules: %v", err))
+		return warnings, errors
+	}
+	mods := strings.Split(string(out), "\n")
+
+	wantModules := sets.NewString()
+	loadModules := sets.NewString()
+	wantModules.Insert(ipvsModules...)
+	loadModules.Insert(mods...)
+	modules := wantModules.Difference(loadModules).UnsortedList()
+
+	// Check builtin modules exist or not
+	if len(modules) != 0 {
+		kernelVersionFile := "/proc/sys/kernel/osrelease"
```

Comment:
Please move all of this kube-proxy specific logic to e.g. `pkg/util/ipvs`. For instance a method called `HostSupportsIPVS()` using an interface to abstract away the OS-specific differences.

Reply:
you are right. But now that we known which `proxy mode` we use from kube-proxy ComponentConfig, Adding `pre-check` when `kubeadm init` master also well. WDYT?